File: mobile/ios/FindlyKit/Sources/FindlyKit/DesignSystem/Tokens/ColorTokens.swift
import SwiftUI

extension Color {
    /// `0xRRGGBB` convenience initializer for the fixed design-token defaults below.
    init(hex: UInt32) {
        let r = Double((hex >> 16) & 0xFF) / 255
        let g = Double((hex >> 8) & 0xFF) / 255
        let b = Double(hex & 0xFF) / 255
        self.init(red: r, green: g, blue: b)
    }
}

/// specs/004-ios-client.md §2.1 — the semantic color vocabulary, identical to the Android client's
/// token names. Components read these fields ONLY, never a literal `Color(...)`.
public struct ColorTokens: Equatable {
    public var primary: Color
    public var onPrimary: Color
    public var secondary: Color
    public var surface: Color
    public var onSurface: Color
    public var surfaceVariant: Color
    public var danger: Color
    public var onDanger: Color
    public var success: Color
    public var warning: Color
    public var outline: Color

    public init(
        primary: Color, onPrimary: Color, secondary: Color, surface: Color, onSurface: Color,
        surfaceVariant: Color, danger: Color, onDanger: Color, success: Color, warning: Color, outline: Color
    ) {
        self.primary = primary
        self.onPrimary = onPrimary
        self.secondary = secondary
        self.surface = surface
        self.onSurface = onSurface
        self.surfaceVariant = surfaceVariant
        self.danger = danger
        self.onDanger = onDanger
        self.success = success
        self.warning = warning
        self.outline = outline
    }

    // Design direction 2a "Ember / Dusk" (design/findly-design-system/2a-ember-dusk/HANDOFF.md,
    // 2026-08-06) — supersedes the 2026-07-20 teal palette. Every text/essential-icon pairing
    // verified WCAG 2.1 AA (ratios in that handoff and specs/004 §2.1).
    public static let light = ColorTokens(
        primary: Color(hex: 0x3A46C8), onPrimary: Color(hex: 0xFFFFFF), secondary: Color(hex: 0x0E7C8F),
        surface: Color(hex: 0xF2F4FB), onSurface: Color(hex: 0x10142A), surfaceVariant: Color(hex: 0xE2E6F5),
        danger: Color(hex: 0xB3261E), onDanger: Color(hex: 0xFFFFFF), success: Color(hex: 0x10714A),
        warning: Color(hex: 0x8A5A00), outline: Color(hex: 0xA9B0CE)
    )

    public static let dark = ColorTokens(
        primary: Color(hex: 0x7C8BFF), onPrimary: Color(hex: 0x0A0F27), secondary: Color(hex: 0x4FE3D0),
        surface: Color(hex: 0x0B0F1C), onSurface: Color(hex: 0xE8ECF7), surfaceVariant: Color(hex: 0x161D33),
        danger: Color(hex: 0xFF6B6B), onDanger: Color(hex: 0x2A0708), success: Color(hex: 0x52E39B),
        warning: Color(hex: 0xFFC44D), outline: Color(hex: 0x3A4463)
    )
}

// design/findly-design-system/2a-ember-dusk/HANDOFF.md — two theme-invariant colors called out by
// name, both contrast traps a naive substitution would get wrong:
public extension Color {
    /// Light `outline` (`#A9B0CE`) is only 2.1:1 — legal for decorative hairlines/dividers only.
    /// Any stroke that carries meaning (unselected control border, focus ring, input outline) uses
    /// this stronger color (3.4:1) instead. In dark, `ColorTokens.dark.outline` itself already
    /// clears 3:1 and may be used for both purposes.
    static let findlyOutlineStrong = Color(hex: 0x6B739A)
    /// The dot inside a `primary` marker bubble's "● NOW" pill — fixed at `#52E39B` in BOTH
    /// themes (5.4:1 on `#3A46C8`). Light-theme `success` (`#10714A`) measures 1.2:1 there and
    /// must never be substituted in.
    static let findlyMarkerOnlineDot = Color(hex: 0x52E39B)
    /// Text drawn on top of `findlyMarkerOnlineDot`.
    static let findlyMarkerOnlineDotOn = Color(hex: 0x062418)
}

File: mobile/android/app/src/main/java/com/findly/android/ui/designsystem/token/ColorTokens.kt
package com.findly.android.ui.designsystem.token

import androidx.compose.runtime.Immutable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color

/**
 * The color half of the design-token contract (specs/003-android-client.md §4.1/§4.2).
 * These are the ONLY color values allowed to exist anywhere in the app outside this file —
 * every component reads [com.findly.android.ui.designsystem.FindlyTheme.colors] instead of
 * constructing a [Color] literal.
 *
 * The 11 fields above the divider are the normative §4.1 vocabulary (do not rename without a
 * spec PR). The fields below it are additive, theme-resolved helper values the 2a "Ember / Dusk"
 * handoff (design/findly-design-system/2a-ember-dusk/HANDOFF.md) needs but that don't belong in
 * the 11-role contract itself — kept on this same per-theme data class (rather than as bare
 * top-level constants, as the handoff's own ready-to-paste `FindlyColors.kt` does) so they
 * resolve automatically with the active theme through the existing [LocalFindlyColors]
 * composition local, the same way every contract color already does. Any translucent tint the
 * handoff derives from `primary` (focus rings, the textfield focus ring, the button shadow tint,
 * the geofence circle fill) is computed at the call site as `primary.copy(alpha = …)` instead of
 * being pinned here as a light-only literal — that keeps it correct in dark automatically, since
 * the handoff itself only ever specifies these as an alpha over the light `primary` hex, which is
 * `primary` itself.
 */
@Immutable
data class FindlyColorTokens(
    val primary: Color,
    val onPrimary: Color,
    val secondary: Color,
    val surface: Color,
    val onSurface: Color,
    val surfaceVariant: Color,
    val danger: Color,
    val onDanger: Color,
    val success: Color,
    val warning: Color,
    val outline: Color,
    // --- Additive, non-contract helpers (2a handoff) ---
    /**
     * Strokes that carry meaning (an unselected control border, a focus ring, an input outline)
     * MUST use this instead of [outline]. Light [outline] is only 2.1:1 — legal for decorative
     * hairlines and dividers only. In dark, [outline] itself already clears 3:1, so this equals
     * [outline] there (HANDOFF.md "Two rules that must survive review", rule 1).
     */
    val outlineStrong: Color,
    /** The dot inside a `primary` [com.findly.android.ui.designsystem.components.FindlyMapMarkerBubble]
     * bubble's "NOW" pill — `#52E39B` in BOTH themes (5.4:1 on `#3A46C8`). Never light-theme
     * [success] there (measures 1.2:1 and disappears — HANDOFF.md rule 2). */
    val markerOnlineDot: Color,
    /** Text color on top of [markerOnlineDot]. */
    val markerOnlineDotOn: Color,
    /** [onSurface] at ~70% — used for muted/secondary text (e.g. `FindlyListRow` subtitles).
     * HANDOFF.md gives exact hex per theme rather than a computed alpha. */
    val subtleText: Color,
    /** Neutral-black shadow opacity for `FindlyTheme.elevation.level1/2/3` (HANDOFF.md's
     * elevation table: `{blur, y, opacity light|dark}`). Shadows stay neutral black in both
     * themes — never tinted — except `FindlyButton`'s primary variant, which tints its `level2`
     * shadow with `primary` on light only (see [isDark] / that component). */
    val shadowLevel1Alpha: Float,
    val shadowLevel2Alpha: Float,
    val shadowLevel3Alpha: Float,
    /** True for [DarkFindlyColors]. The one thing HANDOFF.md specifies as a *light-only*
     * exception (`FindlyButton` primary's tinted shadow) needs to know which theme is active;
     * every other value above already resolves per-theme without a branch. */
    val isDark: Boolean,
)

/**
 * Design 2a — "Ember / Dusk" (design/findly-design-system/2a-ember-dusk/HANDOFF.md).
 * Ember layout/type/spacing/shape; Dusk indigo/cyan color values. Every text/essential-icon
 * pairing meets WCAG 2.1 AA (ratios in HANDOFF.md). Fully swappable via this seam.
 */
val LightFindlyColors = FindlyColorTokens(
    primary = Color(0xFF3A46C8),
    onPrimary = Color(0xFFFFFFFF),
    secondary = Color(0xFF0E7C8F),
    surface = Color(0xFFF2F4FB),
    onSurface = Color(0xFF10142A),
    surfaceVariant = Color(0xFFE2E6F5),
    danger = Color(0xFFB3261E),
    onDanger = Color(0xFFFFFFFF),
    success = Color(0xFF10714A),
    warning = Color(0xFF8A5A00),
    outline = Color(0xFFA9B0CE),
    outlineStrong = Color(0xFF6B739A),
    markerOnlineDot = Color(0xFF52E39B),
    markerOnlineDotOn = Color(0xFF062418),
    subtleText = Color(0xFF4E5675),
    shadowLevel1Alpha = 0.10f,
    shadowLevel2Alpha = 0.14f,
    shadowLevel3Alpha = 0.18f,
    isDark = false,
)

val DarkFindlyColors = FindlyColorTokens(
    primary = Color(0xFF7C8BFF),
    onPrimary = Color(0xFF0A0F27),
    secondary = Color(0xFF4FE3D0),
    surface = Color(0xFF0B0F1C),
    onSurface = Color(0xFFE8ECF7),
    surfaceVariant = Color(0xFF161D33),
    danger = Color(0xFFFF6B6B),
    onDanger = Color(0xFF2A0708),
    success = Color(0xFF52E39B),
    warning = Color(0xFFFFC44D),
    outline = Color(0xFF3A4463),
    outlineStrong = Color(0xFF3A4463),
    markerOnlineDot = Color(0xFF52E39B),
    markerOnlineDotOn = Color(0xFF062418),
    subtleText = Color(0xFF98A1BD),
    shadowLevel1Alpha = 0.30f,
    shadowLevel2Alpha = 0.45f,
    shadowLevel3Alpha = 0.60f,
    isDark = true,
)

/** `FindlyButton` disabled label color, both themes (HANDOFF.md: "Disabled: fill surfaceVariant,
 * label `#8D93AB`") — a fixed value the handoff gives once, not per-theme. */
val ButtonDisabledLabel = Color(0xFF8D93AB)

/** `FindlyTextField` disabled treatment (HANDOFF.md: "Disabled: fill `#E8EAF2`, border `#D3D7E6`,
 * text `#8D93AB`") — fixed values, not theme-resolved (the mocks only specify one disabled
 * treatment, not a light/dark pair). */
val TextFieldDisabledFill = Color(0xFFE8EAF2)
val TextFieldDisabledBorder = Color(0xFFD3D7E6)
val TextFieldDisabledText = Color(0xFF8D93AB)

/** `FindlySwitchRow`'s native M3 switch thumb — literally "white thumb" in both themes/states per
 * HANDOFF.md (`FindlySwitchRow`: "on = `primary` track, off = `outline` track, white thumb"),
 * not derived from any theme token. */
val SwitchThumbColor = Color(0xFFFFFFFF)

val LocalFindlyColors = staticCompositionLocalOf { LightFindlyColors }
